### PR TITLE
feat(core,schemas): koaLogSession middleware

### PR DESCRIPTION
**Pull request by IceHe**
## Summary

- Log session id and application id from OIDC Interaction

## Linear Issue Reference

LOG-2323

## Testing

image

and all existing tests.

### CODE REVIEW

File: packages/schemas/src/types/log.ts [hunk 1]
```diff
@@ -12,38 +12,41 @@ export interface BaseLogPayload {
...
 }
 
-interface CommonFields {
+export interface SessionLogPayload {
+  applicationId?: string;
   sessionId?: string;
 }
 
-interface RegisterUsernamePasswordLogPayload extends CommonFields {
+type ArbitraryLogPayload = Record<string, unknown>;
```
[the remark this answers is not in the document]
Comment (IceHe) [reply packages/schemas/src/types/log.ts:20]: if *LogPayload does not extend it, `deepmerge` cannot pass the type check.

image

maybe better just disable the lint rule here.

File: packages/core/src/middleware/koa-log-session.test.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      logSession,
+      log,
+    };
+
+    await expect(koaLogSession(new Provider(''))(ctx, next)).resolves.not.toThrow();
+    expect(interactionDetails).toHaveBeenCalled();
+    expect(logSession).toHaveBeenCalledWith({ sessionId, applicationId });
+  });
+
+  it('should log session id and application id', async () => {
```
Comment (wangsijie) [inline packages/core/src/middleware/koa-log-session.test.ts:45]: what's the difference between this case and the case in line 33.

Comment (IceHe) [reply packages/core/src/middleware/koa-log-session.test.ts:45]: i should remove `expect(logSession).toHaveBeenCalledWith({ sessionId, applicationId });` from previous case.

updated. PTAL

File: packages/core/src/middleware/koa-log-session.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import { MiddlewareType } from 'koa';
+import { Provider } from 'oidc-provider';
+
+import { WithLogContext } from '@/middleware/koa-log';
+
+export default function koaLogSession<StateT, ContextT extends WithLogContext, ResponseBodyT>(
+  provider: Provider
+): MiddlewareType<StateT, ContextT, ResponseBodyT> {
+  return async (ctx, next) => {
+    const nextPromise = next();
```
Comment (wangsijie) [inline packages/core/src/middleware/koa-log-session.ts:10]: should await here?

File: packages/core/src/middleware/koa-log.ts [hunk 4]
```diff
@@ -60,18 +81,19 @@ const initLogger = (basePayload?: Readonly<BaseLogPayload>) => {
...
-> {
+>(): MiddlewareType<StateT, WithLogContext<ContextT>, ResponseBodyT> {
   return async (ctx, next) => {
     const {
       ip,
       headers: { 'user-agent': userAgent },
     } = ctx.request;
 
     const logger = initLogger({ result: LogResult.Success, ip, userAgent });
+    ctx.logSession = logger.logSession;
```
Comment (wangsijie) [inline packages/core/src/middleware/koa-log.ts:96]: Is it possible to reuse `ctx.log`?

Comment (IceHe) [reply packages/core/src/middleware/koa-log.ts:96]: cannot.
`koa-log-session` middleware does not know using which `type`, the first parameter of `ctx.log(type, payload)`.
seems that we still need to add another function for `koa-log-session` middleware, or `ctx.log` function has to be more complex.

Comment (wangsijie) [reply packages/core/src/middleware/koa-log.ts:96]: add some methods like "addBreadCrumb"? more generic function.

Comment (IceHe) [reply packages/core/src/middleware/koa-log.ts]: refactored: merged SessionLogPayload into BaseLogPayload. PTAL

**Review by demonzoo — APPROVED** [review]
LGTM

**Pull request merged**